aws-crt-http/HttpConnection.cs

```diff
@@ -28,39 +30,135 @@ public enum OutgoingBodyStreamState
         Done = 1,
     }
 
-    public delegate void OnConnectionSetup(int errorCode);
-    public delegate void OnConnectionShutdown(int errorCode);
-    public delegate OutgoingBodyStreamState OnStreamOutgoingBody(HttpClientStream stream, byte[] buffer, out UInt64 bytesWritten);
-    public delegate void OnIncomingHeaders(HttpClientStream stream, HttpHeader[] headers);
-    public delegate void OnIncomingHeaderBlockDone(HttpClientStream stream, bool hasBody);
-    public delegate void OnIncomingBody(HttpClientStream stream, byte[] data, ref UInt64 windowSize);
-    public delegate void OnStreamComplete(HttpClientStream stream, int errorCode);
-
-    internal delegate int OnStreamOutgoingBodyNative(
-        [In, Out, MarshalAs(UnmanagedType.LPArray, SizeParamIndex=1)] byte[] buffer, 
-        UInt64 size,
-        out UInt64 bytesWritten);
-    internal delegate void OnIncomingHeadersNative(
-        Int32 responseCode,
-        [In, MarshalAs(UnmanagedType.LPArray, SizeParamIndex=2)] HttpHeader[] headers, 
-        UInt32 count);
-    internal delegate void OnIncomingHeaderBlockDoneNative(bool hasBody);
-    internal delegate void OnIncomingBodyNative(
-        [In, MarshalAs(UnmanagedType.LPArray, SizeParamIndex=1)] byte[] buffer, 
-        UInt64 size,
-        ref UInt64 windowSize);
-    internal delegate void OnStreamCompleteNative(int errorCode);
+    public class StreamResult
+    {
+        public int ErrorCode { get; internal set; }
+
+        internal StreamResult(int errorCode)
+        {
+            ErrorCode = errorCode;
+        }
+    }
+
+    public abstract class HttpClientStreamEventArgs : EventArgs
+    {
+        public HttpClientStream Stream { get; private set; }
+
+        internal HttpClientStreamEventArgs(HttpClientStream stream)
+        {
+            Stream = stream;
+        }
+    }
+
+    public class IncomingHeadersEventArgs : HttpClientStreamEventArgs
+    {
+        public HttpHeader[] Headers { get; private set; }
+
+        internal IncomingHeadersEventArgs(HttpClientStream stream, HttpHeader[] headers)
+            : base(stream)
+        {
+            Headers = headers;
+        }
+    }
+
+    public class IncomingHeadersDoneEventArgs : HttpClientStreamEventArgs
+    {
+        public bool HasBody { get; private set; }
+
+        internal IncomingHeadersDoneEventArgs(HttpClientStream stream, bool hasBody)
+            : base(stream)
+        {
+            HasBody = hasBody;
+        }
+    }
+
+    public class IncomingBodyEventArgs : HttpClientStreamEventArgs
+    {
+        public byte[] Data { get; private set; }
+        public UInt64 WindowSize { get; private set; }

```

Comment:
maybe not worth exposing WindowSize as a read-only thing?
It's always equal to Data.length...
On the other hand it's just one more variable on the class, user's don't need to look at it

Maybe name it WindowDecrementSize? WindowShrinkSize? WindowLoss?

Reply:
That should have been writable publicly, fixed. Good catch!